File: packages/core/src/mutate/tree.ts
import { sha256 } from "../core/hash.js";

// Mutable document tree for the kernel. Blocks carry id + raw bytes + trivia so
// the tree can be re-rendered by splice (03 §2.2) after ops. This is the
// working representation an apply() builds, mutates, and renders.

export interface MutBlock {
  id: string;
  type: string;
  raw: string;
  trivia: string; // trailing inter-block trivia (top-level tiling)
  attrs: Record<string, unknown>;
  children: MutBlock[];
  /** true if raw was supplied by an op (serialize as-is; 03 §2.2) */
  dirty?: boolean;
}

export interface MutDoc {
  docId: string;
  path: string;
  format: string;
  leadingTrivia: string;
  frontmatterRaw: string | null; // incl. fences; rendered verbatim before blocks
  children: MutBlock[];
}

export class MutationError extends Error {
  code: string;
  data: Record<string, unknown>;
  constructor(code: string, message: string, data: Record<string, unknown> = {}) {
    super(message);
    this.code = code;
    this.data = data;
  }
}

/** Find a block by id anywhere in the doc; returns the block + its parent list. */
export function locate(doc: MutDoc, blockId: string): { block: MutBlock; siblings: MutBlock[]; index: number } | null {
  const search = (list: MutBlock[]): { block: MutBlock; siblings: MutBlock[]; index: number } | null => {
    for (let i = 0; i < list.length; i++) {
      const b = list[i]!;
      if (b.id === blockId) return { block: b, siblings: list, index: i };
      const found = search(b.children);
      if (found) return found;
    }
    return null;
  };
  return search(doc.children);
}

export function rawHashHex(raw: string): string {
  return sha256(raw).toString("hex");
}

/** The block that owns `children` (by reference identity), or null at top level. */
export function ownerOf(doc: MutDoc, children: MutBlock[]): MutBlock | null {
  if (children === doc.children) return null;
  let found: MutBlock | null = null;
  const walk = (list: MutBlock[]): void => {
    for (const b of list) {
      if (b.children === children) { found = b; return; }
      walk(b.children);
    }
  };
  walk(doc.children);
  return found;
}

/**
 * A container's raw is the authoritative render source (03 §2.2), so a
 * structural change to its children (insert/remove/move) must invalidate the
 * container's cached raw. Mark the owning container dirty so renderBlock
 * reconstructs it from the mutated children rather than emitting the now-stale
 * raw verbatim; ancestors see it via hasDirtyDescendant and reconstruct in turn.
 * Top-level siblings (doc.children) need no marking — renderDoc renders each
 * top-level block directly.
 */
export function markContainerDirty(doc: MutDoc, children: MutBlock[]): void {
  const owner = ownerOf(doc, children);
  if (owner) owner.dirty = true;
}

// Render the doc by splice (03 §2.2): leading trivia, optional frontmatter, then
// each top-level block + trivia. A block whose subtree contains no dirty node
// emits its retained raw verbatim. When a descendant is dirty, the block is
// re-rendered from its children (indented to their container) so nested edits
// — e.g. a task checkbox toggle inside a list — surface without disturbing
// untouched siblings.
export function renderDoc(doc: MutDoc): string {
  const out: string[] = [doc.leadingTrivia];
  if (doc.frontmatterRaw !== null) out.push(doc.frontmatterRaw);
  for (const b of doc.children) {
    out.push(renderBlock(b, 0));
    out.push(b.trivia);
  }
  return out.join("");
}

function hasDirtyDescendant(b: MutBlock): boolean {
  if (b.dirty) return true;
  return b.children.some(hasDirtyDescendant);
}

// Render one block. Leaf or fully-clean subtree ⇒ retained raw. Otherwise
// reconstruct from children (a container with a dirty descendant).
function renderBlock(b: MutBlock, depth: number): string {
  if (b.children.length === 0 || !hasDirtyDescendant(b)) return b.raw;
  // A list is reconstructed from its items: each item's raw already carries its
  // own marker + nested content (correctly indented), so we re-mark (ordered
  // lists renumber after a reorder/insert) and join with the list's own
  // tight/loose separator. This matches the canonical renderer, which treats a
  // top-level block's raw as the authoritative source for its whole subtree.
  if (b.type === "list") return renderList(b);
  // Other containers (blockquote/table, or a list_item outside a list): fall
  // back to the child-indent rebuild. Lists are the common mutable container.
  const indent = "  ".repeat(depth);
  return b.children.map((c) => indent + renderBlock(c, depth + 1).replace(/\n/g, "\n" + indent)).join("\n");
}

const MARKER_RE = /^(\s*)([-*+]|\d+[.)])(\s+)/;

// Reconstruct a markdown list from its (possibly reordered / edited / added /
// removed) items. Item raws are self-contained (marker + text + nested); we
// only re-derive the leading marker so an ordered list renumbers correctly.
function renderList(list: MutBlock): string {
  const ordered = Boolean(list.attrs.ordered);
  const start = typeof list.attrs.start === "number" ? list.attrs.start : 1;
  const sep = list.raw.includes("\n\n") ? "\n\n" : "\n";
  return list.children.map((item, i) => renderItem(item, ordered ? `${start + i}. ` : "- ")).join(sep);
}

// Render one list item under the given marker. A retained item (no strictly
// deeper change) keeps its raw, only re-marking the first line so ordered
// numbers follow the current position. An item with a dirty descendant is
// rebuilt from its child blocks, marker on the first line and continuations
// indented by the marker width.
function renderItem(item: MutBlock, marker: string): string {
  if (item.children.some(hasDirtyDescendant)) {
    const body = item.children.map((c) => (c.type === "list" ? renderList(c) : renderBlock(c, 0))).join("\n");
    return applyMarker(body, marker);
  }
  return applyMarker(stripMarker(item.raw), marker);
}

// Remove a leading list marker + its indentation from the first line, and the
// same indent width from continuation lines, yielding the marker-free body.
function stripMarker(raw: string): string {
  const lines = raw.split("\n");
  const m = MARKER_RE.exec(lines[0] ?? "");
  if (!m) return raw;
  const width = m[0].length;
  return lines
    .map((ln, i) => (i === 0 ? ln.slice(width) : ln.slice(0, width).trim() === "" ? ln.slice(width) : ln))
    .join("\n");
}

// Prepend a marker to a marker-free body: marker on the first line, the rest
// indented by the marker width so nested content stays under the item.
function applyMarker(body: string, marker: string): string {
  const pad = " ".repeat(marker.length);
  return body.split("\n").map((ln, i) => (i === 0 ? marker + ln : ln.length ? pad + ln : ln)).join("\n");
}

/** Ordered ids of a parent's direct children (for parent_children_hash CAS). */
export function childIds(list: MutBlock[]): string[] {
  return list.map((b) => b.id);
}

export function parentChildrenHash(list: MutBlock[]): string {
  return sha256(childIds(list).join(",")).toString("hex");
}

File: crates/omgbase-mutate/src/tree.rs
//! The working tree (`spec/mutate/README.md` §1): a document loaded from the
//! store into mutable blocks that carry their id, raw bytes, trailing trivia,
//! attrs, children and a render flag. Blocks are addressed by **path** — the
//! indices from the top level down — because Rust cannot hold a reference to
//! a block while mutating its siblings; `locate` finds the path, the
//! accessors resolve it.

use omgbase_format::hash::{hex, sha256};
use serde_json::{Map, Value};

/// A block of the working tree (§1 `MutBlock`).
#[derive(Clone, Debug, PartialEq)]
pub struct MutBlock {
    pub id: String,
    /// The spec/format §3 kind name (`blocks.type`).
    pub kind: String,
    /// The raw blob's bytes.
    pub raw: String,
    /// The trivia blob's bytes, or `""` when `trivia_hash` was `NULL`.
    pub trivia: String,
    /// The block's attrs as stored (a JSON object).
    pub attrs: Map<String, Value>,
    pub children: Vec<MutBlock>,
    /// The render flag (§3): `false` when loaded; set by the ops whose
    /// children changed.
    pub dirty: bool,
}

impl MutBlock {
    /// A childless, clean block with empty attrs.
    #[must_use]
    pub fn new(id: &str, kind: &str, raw: &str, trivia: &str) -> Self {
        Self {
            id: id.to_owned(),
            kind: kind.to_owned(),
            raw: raw.to_owned(),
            trivia: trivia.to_owned(),
            attrs: Map::new(),
            children: Vec::new(),
            dirty: false,
        }
    }

    /// This block and every descendant, pre-order.
    pub fn iter(&self) -> impl Iterator<Item = &MutBlock> {
        Walk { stack: vec![self] }
    }

    /// Whether this block or any descendant is dirty (§3).
    #[must_use]
    pub fn has_dirty_descendant(&self) -> bool {
        self.dirty || self.children.iter().any(MutBlock::has_dirty_descendant)
    }

    /// Mark this block and its whole subtree clean.
    pub fn mark_subtree_clean(&mut self) {
        self.dirty = false;
        for c in &mut self.children {
            c.mark_subtree_clean();
        }
    }

    /// The ids of this block and every descendant, pre-order.
    #[must_use]
    pub fn subtree_ids(&self) -> Vec<String> {
        self.iter().map(|b| b.id.clone()).collect()
    }
}

struct Walk<'a> {
    stack: Vec<&'a MutBlock>,
}

impl<'a> Iterator for Walk<'a> {
    type Item = &'a MutBlock;

    fn next(&mut self) -> Option<Self::Item> {
        let b = self.stack.pop()?;
        self.stack.extend(b.children.iter().rev());
        Some(b)
    }
}

/// A document in the working tree (§1 `MutDoc`).
#[derive(Clone, Debug, PartialEq)]
pub struct MutDoc {
    pub doc_id: String,
    pub path: String,
    /// `docs.format` (`markdown`).
    pub format: String,
    /// `docs.leading_trivia`.
    pub leading_trivia: String,
    /// The current revision's frontmatter blob + `docs.frontmatter_trivia`,
    /// or `None` when the document has no frontmatter.
    pub frontmatter_raw: Option<String>,
    /// The top-level blocks.
    pub children: Vec<MutBlock>,
}

/// Where a block sits: the indices from the top level down to it. The last
/// element is its index among its siblings; the prefix is its parent's path
/// (empty at the top level).
pub type BlockPath = Vec<usize>;

impl MutDoc {
    /// A document with no frontmatter and the given top-level blocks.
    #[must_use]
    pub fn new(doc_id: &str, path: &str, children: Vec<MutBlock>) -> Self {
        Self {
            doc_id: doc_id.to_owned(),
            path: path.to_owned(),
            format: "markdown".to_owned(),
            leading_trivia: String::new(),
            frontmatter_raw: None,
            children,
        }
    }

    /// Every block at every depth, pre-order.
    pub fn iter(&self) -> impl Iterator<Item = &MutBlock> {
        Walk {
            stack: self.children.iter().rev().collect(),
        }
    }

    /// §1: find a block anywhere in the document; its path.
    #[must_use]
    pub fn locate(&self, block_id: &str) -> Option<BlockPath> {
        fn search(list: &[MutBlock], id: &str, path: &mut BlockPath) -> bool {
            for (i, b) in list.iter().enumerate() {
                path.push(i);
                if b.id == id || search(&b.children, id, path) {
                    return true;
                }
                path.pop();
            }
            false
        }
        let mut path = Vec::new();
        search(&self.children, block_id, &mut path).then_some(path)
    }

    /// Whether `block_id` is anywhere in the document.
    #[must_use]
    pub fn contains(&self, block_id: &str) -> bool {
        self.locate(block_id).is_some()
    }

    /// The sibling list a parent path names (`[]` → the top level).
    #[must_use]
    pub fn siblings(&self, parent: &[usize]) -> &Vec<MutBlock> {
        let mut list = &self.children;
        for &i in parent {
            list = &list[i].children;
        }
        list
    }

    /// Mutable [`MutDoc::siblings`].
    pub fn siblings_mut(&mut self, parent: &[usize]) -> &mut Vec<MutBlock> {
        let mut list = &mut self.children;
        for &i in parent {
            list = &mut list[i].children;
        }
        list
    }

    /// The block at `path`.
    ///
    /// # Panics
    ///
    /// If `path` is empty or does not name a block.
    #[must_use]
    pub fn block(&self, path: &[usize]) -> &MutBlock {
        let (last, parent) = path.split_last().expect("a block path is non-empty");
        &self.siblings(parent)[*last]
    }

    /// Mutable [`MutDoc::block`].
    ///
    /// # Panics
    ///
    /// If `path` is empty or does not name a block.
    pub fn block_mut(&mut self, path: &[usize]) -> &mut MutBlock {
        let (last, parent) = path.split_last().expect("a block path is non-empty");
        &mut self.siblings_mut(parent)[*last]
    }

    /// §1 `owner_of`: the block owning the sibling list at `parent`, or
    /// `None` at the top level.
    #[must_use]
    pub fn owner_of(&self, parent: &[usize]) -> Option<&MutBlock> {
        if parent.is_empty() {
            None
        } else {
            Some(self.block(parent))
        }
    }

    /// §1 `mark_container_dirty`: a structural change to a sibling list
    /// invalidates its owner's raw; the top level needs no marking.
    pub fn mark_container_dirty(&mut self, parent: &[usize]) {
        if !parent.is_empty() {
            self.block_mut(parent).dirty = true;
        }
    }

    /// The ids of every block, pre-order.
    #[must_use]
    pub fn all_ids(&self) -> Vec<String> {
        self.iter().map(|b| b.id.clone()).collect()
    }
}

/// `hex(sha256(raw))`: the content CAS token (§1.2).
#[must_use]
pub fn raw_hash_hex(raw: &str) -> String {
    hex(&sha256(raw.as_bytes()))
}

/// The ordered ids of a sibling list.
#[must_use]
pub fn child_ids(list: &[MutBlock]) -> Vec<String> {
    list.iter().map(|b| b.id.clone()).collect()
}

/// §1.2: `hex(sha256(child ids joined by ","))`.
#[must_use]
pub fn parent_children_hash(list: &[MutBlock]) -> String {
    hex(&sha256(child_ids(list).join(",").as_bytes()))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn nested() -> MutDoc {
        let mut list = MutBlock::new("l", "list", "- a\n- b", "\n");
        let mut a = MutBlock::new("a", "list_item", "- a", "");
        a.children.push(MutBlock::new("p", "paragraph", "a", ""));
        list.children.push(a);
        list.children
            .push(MutBlock::new("b", "list_item", "- b", ""));
        MutDoc::new(
            "d_0",
            "a.md",
            vec![MutBlock::new("h", "heading", "# H", "\n\n"), list],
        )
    }

    #[test]
    fn locate_returns_paths_and_accessors_resolve_them() {
        let doc = nested();
        assert_eq!(doc.locate("h"), Some(vec![0]));
        assert_eq!(doc.locate("l"), Some(vec![1]));
        assert_eq!(doc.locate("b"), Some(vec![1, 1]));
        assert_eq!(doc.locate("p"), Some(vec![1, 0, 0]));
        assert_eq!(doc.locate("zz"), None);
        assert_eq!(doc.block(&[1, 0, 0]).raw, "a");
        assert_eq!(doc.siblings(&[1]).len(), 2);
        assert_eq!(doc.owner_of(&[1]).map(|b| b.id.as_str()), Some("l"));
        assert!(doc.owner_of(&[]).is_none());
        assert_eq!(doc.all_ids(), ["h", "l", "a", "p", "b"]);
    }

    #[test]
    fn dirty_marks_propagate_to_ancestors_only_by_query() {
        let mut doc = nested();
        assert!(!doc.block(&[1]).has_dirty_descendant());
        doc.mark_container_dirty(&[1, 0]);
        assert!(doc.block(&[1, 0]).dirty);
        assert!(doc.block(&[1]).has_dirty_descendant());
        assert!(!doc.block(&[1]).dirty);
        doc.mark_container_dirty(&[]);
        doc.block_mut(&[1]).mark_subtree_clean();
        assert!(!doc.block(&[1]).has_dirty_descendant());
    }

    #[test]
    fn hashes() {
        assert_eq!(
            raw_hash_hex("abc"),
            "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad"
        );
        let doc = nested();
        assert_eq!(child_ids(&doc.children), ["h", "l"]);
        assert_eq!(parent_children_hash(&doc.children), raw_hash_hex("h,l"));
        assert_eq!(parent_children_hash(&[]), raw_hash_hex(""));
    }
}
